Increase stability of builds and tests on CI

We have a lot of red builds from the master branch and PRs on CI, we should investigate and ensure a more stable builds and tests.

# Examples of recent instabilities

## Build 1036 (React Native iOS build failures)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1036/pipeline

<details>
  <summary>Log</summary>

```

node /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/node_modules/react-native/cli.js bundle --entry-file index.ios.js --platform ios --dev false --reset-cache --bundle-output /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle --assets-dest /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app
warning: the transform cache was reset.
Loading dependency graph, done.
error Unable to resolve module `url` from `/Users/realm/workspace/realm_realm-js_master/tests/react-test-app/node_modules/realm-tests/session-tests.js`: Module `url` does not exist in the Haste module map

This might be related to https://github.com/facebook/react-native/issues/4968
To resolve try the following:
  1. Clear watchman watches: `watchman watch-del-all`.
  2. Delete the `node_modules` folder: `rm -rf node_modules && npm install`.
  3. Reset Metro Bundler cache: `rm -rf /tmp/metro-bundler-cache-*` or `npm start -- --reset-cache`.
  4. Remove haste cache: `rm -rf /tmp/haste-map-react-native-packager-*`.. Run CLI with --verbose flag for more details.
+ [[ false != true ]]
+ [[ ! -f /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle ]]
+ echo 'error: File /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle does not exist. This must be a bug with'
error: File /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle does not exist. This must be a bug with
+ echo 'React Native, please report it here: https://github.com/facebook/react-native/issues'
React Native, please report it here: https://github.com/facebook/react-native/issues
+ exit 2

GenerateDSYMFile /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app.dSYM /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/ReactTests (in target 'ReactTestApp' from project 'ReactTests')
    cd /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios
    /Applications/Xcode-11.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/dsymutil /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/ReactTests -o /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app.dSYM

warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 7.0, but the range of supported deployment target versions is 8.0 to 13.2.99. (in target 'RNFS' from project 'Pods')
** TEST BUILD FAILED **
```
  </summary>
</details>

## Build 1035 (React Native iOS build failures)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1035/pipeline

<details>
  <summary>Log</summary>

```
node /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/node_modules/react-native/cli.js bundle --entry-file index.ios.js --platform ios --dev false --reset-cache --bundle-output /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle --assets-dest /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app
warning: the transform cache was reset.
Loading dependency graph, done.
error Unable to resolve module `url` from `/Users/realm/workspace/realm_realm-js_master/tests/react-test-app/node_modules/realm-tests/session-tests.js`: Module `url` does not exist in the Haste module map

This might be related to https://github.com/facebook/react-native/issues/4968
To resolve try the following:
  1. Clear watchman watches: `watchman watch-del-all`.
  2. Delete the `node_modules` folder: `rm -rf node_modules && npm install`.
  3. Reset Metro Bundler cache: `rm -rf /tmp/metro-bundler-cache-*` or `npm start -- --reset-cache`.
  4. Remove haste cache: `rm -rf /tmp/haste-map-react-native-packager-*`.. Run CLI with --verbose flag for more details.
+ [[ false != true ]]
+ [[ ! -f /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle ]]
+ echo 'error: File /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle does not exist. This must be a bug with'
error: File /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/main.jsbundle does not exist. This must be a bug with
+ echo 'React Native, please report it here: https://github.com/facebook/react-native/issues'
React Native, please report it here: https://github.com/facebook/react-native/issues
+ exit 2

GenerateDSYMFile /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app.dSYM /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/ReactTests (in target 'ReactTestApp' from project 'ReactTests')
    cd /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios
    /Applications/Xcode-11.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/dsymutil /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app/ReactTests -o /Users/realm/workspace/realm_realm-js_master/tests/react-test-app/ios/build/Build/Products/Release-iphonesimulator/ReactTests.app.dSYM

warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 7.0, but the range of supported deployment target versions is 8.0 to 13.2.99. (in target 'RNFS' from project 'Pods')

** TEST BUILD FAILED **
```
  </summary>
</details>

## Build 1034 (Gradle related Android build failure)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1034/pipeline/

<details>
  <summary>Log</summary>

```
+ cd react-native/android
+ ./gradlew publishAndroid

Starting a Gradle Daemon, 5 busy and 2 incompatible and 5 stopped Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type ScriptPluginFactory using BuildScopeServices.createScriptPluginFactory().
> Could not create service of type FileAccessTimeJournal using GradleUserHomeScopeServices.createFileAccessTimeJournal().

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 3s
script returned exit code 1
```

  </summary>
</details>

## Build 1033 (Gradle related Android build failure)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1033/pipeline/

<details>
  <summary>Log</summary>

```
+ cd react-native/android
+ ./gradlew publishAndroid
Starting a Gradle Daemon, 6 busy and 3 stopped Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type ScriptPluginFactory using BuildScopeServices.createScriptPluginFactory().
> Could not create service of type ResourceSnapshotterCacheService using GradleUserHomeScopeServices.createResourceSnapshotterCacheService().

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org

BUILD FAILED in 1m 2s
script returned exit code 1
```

  </summary>
</details>

## Build 1032 (Gradle related Android build failure)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1032/pipeline/

<details>
  <summary>Log</summary>

```
+ cd react-native/android
+ ./gradlew publishAndroid
Starting a Gradle Daemon, 5 busy and 1 incompatible and 3 stopped Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type ScriptPluginFactory using BuildScopeServices.createScriptPluginFactory().
> Could not create service of type FileAccessTimeJournal using GradleUserHomeScopeServices.createFileAccessTimeJournal().

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 3s
script returned exit code 1
```

  </summary>
</details>

## Build 1031 (Gradle related Android build failure)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1031/pipeline/

<details>
  <summary>Log</summary>

```
+ cd react-native/android
+ ./gradlew publishAndroid
Starting a Gradle Daemon, 5 busy and 1 incompatible and 1 stopped Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type ScriptPluginFactory using BuildScopeServices.createScriptPluginFactory().
> Could not create service of type FileAccessTimeJournal using GradleUserHomeScopeServices.createFileAccessTimeJournal().

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 2s
script returned exit code 1
```

  </summary>
</details>

## Build 1030 (Jenkins related JNLP4-connection error for MacOS nodes)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1030/pipeline/

Probable root cause might be the VPN between our Jenkins master and the network of the MacOS nodes going offline.

<details>
  <summary>Log</summary>

```
Also:   <cycle to hudson.model.Computer$TerminationRequest: Termination requested at Thu Jan 16 19:52:59 UTC 2020 by Thread[Handling POST /computer/lv_host10_1/doDisconnect from 207.254.61.171 : qtp1146743572-5667014,5,main] [id=5667014] from HTTP request for https://ci.realm.io/computer/lv_host10_1/doDisconnect>
	Caused: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on JNLP4-connect connection from 10.60.220.42/10.60.220.42:8338 failed. The channel is closing down or has closed down
Caused: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on JNLP4-connect connection from 10.60.220.42/10.60.220.42:8338 failed. The channel is closing down or has closed down
	at hudson.remoting.Channel.call(Channel.java:948)
	at hudson.EnvVars.getRemote(EnvVars.java:426)
	at hudson.model.Computer.getEnvironment(Computer.java:1192)
	at org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution$PlaceholderTask$PlaceholderExecutable.run(ExecutorStepExecution.java:751)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
```

  </summary>
</details>

## Build 1029 (Failing unit tests)

https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-js/detail/master/1029/pipeline/

Probable root cause is an actual error, which was probably fixed.

<details>
  <summary>Log</summary>

A lot of nodes have logs similar to:

```
SessionTests : testClientResyncDiscard
   Failed: Identifier 'url' has already been declared
```

  </summary>
</details>

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -215,13 +215,16 @@ def reactNativeIntegrationTests(targetPlatform) {
   def nvm
   if (targetPlatform == "android") {
     nvm = ""
-  }
-  else {
+  } else {
     nvm = "${env.WORKSPACE}/scripts/nvm-wrapper.sh ${nodeVersion}"
   }
 
   dir('integration-tests') {
-    unstash 'android'
+    if (targetPlatform == "android") {
+      unstash 'android'
+    } else {
+      sh "${nvm} npm pack .."
+    }
     // Renaming the package to avoid having to specify version in the apps package.json
     sh 'mv realm-*.tgz realm.tgz'
     // Package up the integration tests

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -54,6 +54,7 @@ stage('check') {
   }
 }
 
+/*
 stage('pretest') {
   parallelExecutors = [:]
     parallelExecutors["eslint"] = testLinux("eslint-ci Release ${nodeTestVersion}", { // "Release" is not used
@@ -127,21 +128,26 @@ stage('test') {
   //}),
   parallel parallelExecutors
 }
+*/
 
 stage('integration tests') {
   parallel(
+    /*
     'React Native on Android':  inAndroidContainer {
       // Locking the Android device to prevent other jobs from interfering
       lock("${NODE_NAME}-android") {
         reactNativeIntegrationTests('android')
       }
     },
+    */
     'React Native on iOS':      buildMacOS { reactNativeIntegrationTests('ios') },
+    /*
     'Electron on Mac':          buildMacOS { electronIntegrationTests('4.1.4', it) },
     'Electron on Linux':        buildLinux { electronIntegrationTests('4.1.4', it) },
     'Node.js v10 on Mac':       buildMacOS { nodeIntegrationTests('10.15.1', it) },
     'Node.js v8 on Linux':      buildLinux { nodeIntegrationTests('8.15.0', it) },
     'Node.js v10 on Linux':     buildLinux { nodeIntegrationTests('10.15.1', it) }
+    */
   )
 }
 

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -381,17 +381,21 @@ def inAndroidContainer(workerFunction) {
         image = buildDockerEnv('ci/realm-js:android-build', '-f Dockerfile.android')
       }
       sh "bash ./scripts/utils.sh set-version ${dependencies.VERSION}"
-      image.inside(
-        // Mounting ~/.android/adbkey(.pub) to reuse the adb keys
-        "-v ${HOME}/.android/adbkey:/home/jenkins/.android/adbkey:ro -v ${HOME}/.android/adbkey.pub:/home/jenkins/.android/adbkey.pub:ro " +
-        // Mounting ~/gradle-cache as ~/.gradle to prevent gradle from being redownloaded
-        "-v ${HOME}/gradle-cache:/home/jenkins/.gradle " +
-        // Mounting ~/ccache as ~/.ccache to reuse the cache across builds
-        "-v ${HOME}/ccache:/home/jenkins/.ccache " +
-        // Mounting /dev/bus/usb with --privileged to allow connecting to the device via USB
-        "-v /dev/bus/usb:/dev/bus/usb --privileged"
-      ) {
-        workerFunction()
+      // Locking on the "android" lock to prevent concurrent usage of the gradle-cache
+      // @see https://github.com/realm/realm-java/blob/master/Jenkinsfile#L65
+      lock("${env.NODE_NAME}-android") {
+        image.inside(
+          // Mounting ~/.android/adbkey(.pub) to reuse the adb keys
+          "-v ${HOME}/.android/adbkey:/home/jenkins/.android/adbkey:ro -v ${HOME}/.android/adbkey.pub:/home/jenkins/.android/adbkey.pub:ro " +
+          // Mounting ~/gradle-cache as ~/.gradle to prevent gradle from being redownloaded
+          "-v ${HOME}/gradle-cache:/home/jenkins/.gradle " +
+          // Mounting ~/ccache as ~/.ccache to reuse the cache across builds
+          "-v ${HOME}/ccache:/home/jenkins/.ccache " +
+          // Mounting /dev/bus/usb with --privileged to allow connecting to the device via USB
+          "-v /dev/bus/usb:/dev/bus/usb --privileged"
+        ) {
+          workerFunction()
+        }
       }
     }
   }

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -54,7 +54,6 @@ stage('check') {
   }
 }
 
-/*
 stage('pretest') {
   parallelExecutors = [:]
     parallelExecutors["eslint"] = testLinux("eslint-ci Release ${nodeTestVersion}", { // "Release" is not used
@@ -128,26 +127,21 @@ stage('test') {
   //}),
   parallel parallelExecutors
 }
-*/
 
 stage('integration tests') {
   parallel(
-    /*
     'React Native on Android':  inAndroidContainer {
       // Locking the Android device to prevent other jobs from interfering
       lock("${NODE_NAME}-android") {
         reactNativeIntegrationTests('android')
       }
     },
-    */
     'React Native on iOS':      buildMacOS { reactNativeIntegrationTests('ios') },
-    /*
     'Electron on Mac':          buildMacOS { electronIntegrationTests('4.1.4', it) },
     'Electron on Linux':        buildLinux { electronIntegrationTests('4.1.4', it) },
     'Node.js v10 on Mac':       buildMacOS { nodeIntegrationTests('10.15.1', it) },
     'Node.js v8 on Linux':      buildLinux { nodeIntegrationTests('8.15.0', it) },
     'Node.js v10 on Linux':     buildLinux { nodeIntegrationTests('10.15.1', it) }
-    */
   )
 }
 

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -381,7 +381,7 @@ def inAndroidContainer(workerFunction) {
       }
       sh "bash ./scripts/utils.sh set-version ${dependencies.VERSION}"
       // Locking on the "android" lock to prevent concurrent usage of the gradle-cache
-      // @see https://github.com/realm/realm-java/blob/master/Jenkinsfile#L65
+      // @see https://github.com/realm/realm-java/blob/00698d1/Jenkinsfile#L65
       lock("${env.NODE_NAME}-android") {
         image.inside(
           // Mounting ~/.android/adbkey(.pub) to reuse the adb keys

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -223,6 +223,7 @@ def reactNativeIntegrationTests(targetPlatform) {
     if (targetPlatform == "android") {
       unstash 'android'
     } else {
+      // Pack up Realm JS into a .tar
       sh "${nvm} npm pack .."
     }
     // Renaming the package to avoid having to specify version in the apps package.json
@@ -242,7 +243,7 @@ def reactNativeIntegrationTests(targetPlatform) {
       sh 'adb uninstall io.realm.tests.reactnative || true' // '|| true' because the app might already not be installed
     } else if (targetPlatform == "ios") {
       dir('ios') {
-        sh 'pod install'
+        sh 'pod install --verbose'
       }
     }
 

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -130,12 +130,7 @@ stage('test') {
 
 stage('integration tests') {
   parallel(
-    'React Native on Android':  inAndroidContainer {
-      // Locking the Android device to prevent other jobs from interfering
-      lock("${NODE_NAME}-android") {
-        reactNativeIntegrationTests('android')
-      }
-    },
+    'React Native on Android':  inAndroidContainer { reactNativeIntegrationTests('android') },
     'React Native on iOS':      buildMacOS { reactNativeIntegrationTests('ios') },
     'Electron on Mac':          buildMacOS { electronIntegrationTests('4.1.4', it) },
     'Electron on Linux':        buildLinux { electronIntegrationTests('4.1.4', it) },

--- a/integration-tests/environments/react-native/harness/runner.js
+++ b/integration-tests/environments/react-native/harness/runner.js
@@ -114,6 +114,8 @@ async function runApp(platform, junitFilePath) {
         const deviceName = "realm-js-integration-tests";
         ensureSimulator(deviceName, 'com.apple.CoreSimulator.SimDeviceType.iPhone-11');
         console.log("Simulator is ready 🚀");
+        // Open the bundle URL (ensures that the simulator can connect to the Metro server, when the app loads)
+        xcode.simctl.openUrl(deviceName, 'http://localhost:8081');
         // Ask React Native to run the ios app
         rn.sync("run-ios", "--no-packager", "--simulator", deviceName);
     } else {

--- a/integration-tests/environments/react-native/harness/runner.js
+++ b/integration-tests/environments/react-native/harness/runner.js
@@ -23,51 +23,67 @@ const rn = require("./react-native-cli");
 const android = require("./android-cli");
 const xcode = require("./xcode-cli");
 
+const IOS_DEVICE_NAME = "realm-js-integration-tests";
+const IOS_DEVICE_TYPE_ID = "com.apple.CoreSimulator.SimDeviceType.iPhone-11";
+
 /**
- * Ensure the simulator is created and booted
+ * Ensure a simulator is created and booted
  */
-function ensureSimulator(deviceName, deviceTypeId) {
-    const version = xcode.xcrun("--version").stdout.trim();
-    console.log(`Using ${version}`);
+function ensureSimulator(platform) {
+    if (platform === "android") {
+        const devices = android.adb.devices();
+        const activeDevices = devices.filter(({ state }) => state === "device");
+        if (activeDevices.length === 0) {
+            throw new Error("Missing an active device: Attach a device via USB or start an emulator");
+        } else {
+            // Ensure the device can access the mocha remote server
+            android.adb.reverseServerPort(MochaRemoteServer.DEFAULT_CONFIG.port);
+        }
+    } else if (platform === "ios") {
 
-    // Determine if the system supports the device type id
-    const { devicetypes } = xcode.simctl.list('devicetypes');
-    const deviceType = devicetypes.find(({ identifier }) => identifier === deviceTypeId);
-    if (!deviceType) {
-        throw new Error(`System doesn't have the "${deviceTypeId}" device type`);
+        const version = xcode.xcrun("--version").stdout.trim();
+        console.log(`Using ${version}`);
+    
+        // Determine if the system supports the device type id
+        const { devicetypes } = xcode.simctl.list('devicetypes');
+        const deviceType = devicetypes.find(({ identifier }) => identifier === IOS_DEVICE_TYPE_ID);
+        if (!deviceType) {
+            throw new Error(`System doesn't have the "${IOS_DEVICE_TYPE_ID}" device type`);
+        }
+    
+        // Shutdown all booted simulators (as they might interfeer by loading and executing the Metro bundle)
+        xcode.simctl.shutdown('all');
+    
+        // Determine if the device exists and has the correct IOS_DEVICE_TYPE_ID and latest runtime
+        const { devices: devicesByType } = xcode.simctl.list('devices', 'available');
+        const availableDevices = [].concat(...Object.keys(devicesByType).map(type => devicesByType[type]));
+        // Filter devices, so we're only focussing on devices of the expected name
+        const devices = availableDevices.filter(({ name }) => name === IOS_DEVICE_NAME);
+    
+        // Delete any existing devices with the expected name
+        for (const device of devices) {
+            console.log(`Deleting simulator (id = ${device.udid})`);
+            xcode.simctl.delete(device.udid);
+        }
+    
+        const { runtimes } = xcode.simctl.list('runtimes', 'ios');
+        const [ runtime ] = runtimes.filter(runtime => runtime.isAvailable);
+        if (!runtime) {
+            throw new Error("No available iOS runtimes");
+        }
+        // Create the device
+        const { stdout } = xcode.simctl.create(IOS_DEVICE_NAME, IOS_DEVICE_TYPE_ID, runtime.identifier);
+        const deviceId = stdout.trim();
+        console.log(`Created simulator device (id = ${deviceId})`);
+        // Boot up the device
+        console.log('Booting simulator');
+        xcode.simctl.boot(deviceId);
+        console.log('Simulator is booting');
+        xcode.simctl.bootstatus(deviceId);
+        console.log("Simulator is ready 🚀");
+    } else {
+        throw new Error(`Unexpected platform: '${platform}'`);
     }
-
-    // Shutdown all booted simulators (as they might interfeer by loading and executing the Metro bundle)
-    xcode.simctl.shutdown('all');
-
-    // Determine if the device exists and has the correct deviceTypeId and latest runtime
-    const { devices: devicesByType } = xcode.simctl.list('devices', 'available');
-    const availableDevices = [].concat(...Object.keys(devicesByType).map(type => devicesByType[type]));
-    // Filter devices, so we're only focussing on devices of the expected name
-    const devices = availableDevices.filter(({ name }) => name === deviceName);
-
-    // Delete any existing devices with the expected name
-    for (const device of devices) {
-        console.log(`Deleting simulator (id = ${device.udid})`);
-        xcode.simctl.delete(device.udid);
-    }
-
-    const { runtimes } = xcode.simctl.list('runtimes', 'ios');
-    const [ runtime ] = runtimes.filter(runtime => runtime.isAvailable);
-    if (!runtime) {
-        throw new Error("No available iOS runtimes");
-    }
-    // Create the device
-    const { stdout } = xcode.simctl.create(deviceName, deviceTypeId, runtime.identifier);
-    const deviceId = stdout.trim();
-    console.log(`Created simulator device (id = ${deviceId})`);
-    // Boot up the device
-    console.log('Booting simulator');
-    xcode.simctl.boot(deviceId);
-    console.log('Simulator is booting');
-    xcode.simctl.bootstatus(deviceId);
-    console.log('Simulator is booted');
-    return deviceId;
 }
 
 async function runApp(platform, junitFilePath) {
@@ -101,34 +117,34 @@ async function runApp(platform, junitFilePath) {
         }
     });
 
+    // Ensure the simulator is booted and ready for the app to start
+    ensureSimulator(platform);
+
     if (platform === "android") {
-        const devices = android.adb.devices();
-        const activeDevices = devices.filter(({ state }) => state === "device");
-        if (activeDevices.length === 0) {
-            throw new Error("Missing an active device: Attach a device via USB or start an emulator");
-        } else {
-            // Ensure the device can access the mocha remote server
-            android.adb.reverseServerPort(MochaRemoteServer.DEFAULT_CONFIG.port);
-        }
         // Ask React Native to run the android app
         rn.sync("run-android", "--no-packager");
     } else if (platform === "ios") {
-        const deviceName = "realm-js-integration-tests";
-        ensureSimulator(deviceName, 'com.apple.CoreSimulator.SimDeviceType.iPhone-11');
-        console.log("Simulator is ready 🚀");
-        // Open the bundle URL (ensures that the simulator can connect to the Metro server, when the app loads)
-        xcode.simctl.openUrl(deviceName, 'http://localhost:8081');
         // Ask React Native to run the ios app
-        rn.sync("run-ios", "--no-packager", "--simulator", deviceName);
+        rn.sync("run-ios", "--no-packager", "--simulator", IOS_DEVICE_NAME);
     } else {
         throw new Error(`Unexpected platform: '${platform}'`);
     }
+
+    // Set an interval that calls "react-native run-ios" every minute to make it refetch the bundle if it fails
+    const retryInterval = setInterval(() => {
+        if (platform === "ios") {
+            // Ask React Native to re-run the ios app
+            rn.sync("run-ios", "--no-packager", "--simulator", IOS_DEVICE_NAME);
+        }
+    }, 60000);
 
     // Run tests with a 5 minute timeout
     return timeout(new Promise((resolve) => {
         console.log("Running tests 🏃‍♂️");
         server.run(resolve);
-    }), 60000 * 5);
+    }), 60000 * 5).finally(() => {
+        clearInterval(retryInterval);
+    });
 }
 
 async function run() {

--- a/integration-tests/environments/react-native/harness/runner.js
+++ b/integration-tests/environments/react-native/harness/runner.js
@@ -120,6 +120,7 @@ async function runApp(platform, junitFilePath) {
     // Ensure the simulator is booted and ready for the app to start
     ensureSimulator(platform);
 
+    console.log("Starting the app");
     if (platform === "android") {
         // Ask React Native to run the android app
         rn.sync("run-android", "--no-packager");
@@ -133,6 +134,7 @@ async function runApp(platform, junitFilePath) {
     // Set an interval that calls "react-native run-ios" every minute to make it refetch the bundle if it fails
     const retryInterval = setInterval(() => {
         if (platform === "ios") {
+            console.log("Retrying starting the iOS app");
             // Ask React Native to re-run the ios app
             rn.sync("run-ios", "--no-packager", "--simulator", IOS_DEVICE_NAME);
         }

--- a/integration-tests/environments/react-native/harness/runner.js
+++ b/integration-tests/environments/react-native/harness/runner.js
@@ -62,9 +62,11 @@ function ensureSimulator(deviceName, deviceTypeId) {
     const deviceId = stdout.trim();
     console.log(`Created simulator device (id = ${deviceId})`);
     // Boot up the device
-    console.log(`Booting simulator`);
+    console.log('Booting simulator');
     xcode.simctl.boot(deviceId);
-    console.log(`Simulator booted`);
+    console.log('Simulator is booting');
+    xcode.simctl.bootstatus(deviceId);
+    console.log('Simulator is booted');
     return deviceId;
 }
 

--- a/integration-tests/environments/react-native/harness/xcode-cli.js
+++ b/integration-tests/environments/react-native/harness/xcode-cli.js
@@ -95,6 +95,15 @@ function launch(device, appBundleIdentifier) {
     return simctl("launch", device, appBundleIdentifier);
 }
 
+/**
+ * Opens a URL in the device browser
+ * @param {string} device 
+ * @param {string} url 
+ */
+function openUrl(device, url) {
+    return simctl("openurl", device, url);
+}
+
 module.exports = {
     xcrun,
     simctl: {
@@ -106,5 +115,6 @@ module.exports = {
         shutdown,
         terminate,
         launch,
+        openUrl,
     }
 };

--- a/integration-tests/environments/react-native/harness/xcode-cli.js
+++ b/integration-tests/environments/react-native/harness/xcode-cli.js
@@ -104,6 +104,14 @@ function openUrl(device, url) {
     return simctl("openurl", device, url);
 }
 
+/**
+ * An undocumented method that blocks until device is booted.
+ * @param {string} device Device to wait for
+ */
+function bootstatus(device) {
+    return simctl("bootstatus", device);
+}
+
 module.exports = {
     xcrun,
     simctl: {
@@ -116,5 +124,6 @@ module.exports = {
         terminate,
         launch,
         openUrl,
+        bootstatus,
     }
 };

